## Eddie: The University of Edinburgh's high performance computer

As well as this resource there is also information on the [University's wiki page.](https://www.wiki.ed.ac.uk/display/ResearchServices/Eddie)

#### What exactly is Eddie?
A computer cluster with 7000+ cores with up to 3TB of memory. 
It uses an Open Grid Scheduler batch system on Scientific Linux 7. Although that seems a bit like too much information sometimes it can be useful to know that when you're trying to troubleshoot certain problems.

___________________________________

### Amelia's tips for using Eddie:
These contain things that took me hours of googling to figure out how to do, so to save anyone else going through that long, frustrating process I've written them all up here. Some of them are personal preferences in how I've set up using Eddie on my Mac but there's a lot of general things too.

#### Table of Contents
1. [Connecting to Eddie](#Connecting-to-Eddie)
2. [Storing data on Eddie and getting files to/from Datastore ](#storing-data-on-eddie-and-getting-files-tofrom-datastore)
3. [Submitting jobs on Eddie](#Submitting-jobs-on-Eddie)
    * [Troubleshooting when your job doesn't work](#troubleshooting-when-your-job-doesnt-work)
4. [Advanced Tips](#advanced-tips)
5. [Some general useful Linux commands](#some-general-useful-linux-commands)
6. [Useful links](#useful-links)


#### Connecting to Eddie
I use Mac OSX so to connect to Eddie I open a bash terminal and type:
```
ssh s1211670@eddie.ecdf.ed.ac.uk
```

* *ssh* is a command that allows us to securely connect to a remote server. 
* *s1211670* is where you put your UUN.
* *@eddie.ecdf.ed.ac.uk* is the name of the remote server.

Additionally, you can use the `-X` or `-Y` flags to forward display to your local computer. **In simple terms you need to do this if you want to plot things** eg. in R using [XQuartz](https://www.xquartz.org/) that's downloaded on your local computer:

```
ssh -Y s1211670@ed.ac.uk
```

If you want to see what other flags you can pass to this command type
```
ssh -h
```
**TIP: In general a command followed by `-h` should offer some help on what arguments can be passed to the command.**

#### Storing data on Eddie and getting files to/from Datastore 
The datastore and eddie servers are separate servers, so if you save something on one it won't be on the other.
###### Datastore storage

* Backed up more so it's generally better for long term storage of data. 
* Group file path: `igmm/datastore/GenScotDepression/`
* I think the general consensus is to create a folder with your name on in `igmm/datastore/GenScotDepression/users/` to keep this shared space tidy.

###### Eddie storage

* You have a personal storage area on Eddie which is called your scratch directory. It can be found at `home/<UUN>` eg:
    ```
    cd /home/s1211670
    ```
    This is cleared every month or so, so not really ideal for long term storage but good for short term. Generally I don't tend to use this space, but I'm not sure what others do, perhaps I am missing a trick!
* Group file path: `/exports/igmm/eddie/GenScotDepression/`
* There is less storage space on this shared eddie space than in datastore, so just bear that in mind and if you have any large files that you have finished using for analysis it's probably good practice to remove them from Eddie and keep a copy on datastore instead.

###### Copy files from datastore to eddie

1. Log into an interactive staging node on eddie. 
```
qlogin -q staging
```
2. Connect to the sftp server using port 22222 and enter your university account password when prompted.*I would love to know if and how to use an SSH key here (more on that later!).*
```
sftp -P 22222 s1211670@cmvm.datastore.ed.ac.uk
```
3. Change your directory on your eddie space to where you want to put your files. eg.:
```
lcd /exports/igmm/eddie/GenScotDepression/amelia/PRS_project/Cross-disorder/Output
lls
```
4. Change directory on datastore space to where the files are located that you want to copy.
```
cd igmm/datastore/GenScotDepression/users/amelia/PRS_project/Cross-disorder/Output
ls
```
5. Copy files or folders across. The `-R` flag indicates we want to copy a directory rather than a single file.
```
get -R significant_snps/
```
6. exit sftp and exit the staging node.
```
exit
exit
```

###### Copy files from eddie to datastore

This is very similar to above but we use `put` instead of `get`.
eg:
```
qlogin -q staging -l h='node2c11'
sftp -P 22222 s1211670@cmvm.datastore.ed.ac.uk

lcd /exports/igmm/eddie/GenScotDepression/amelia/PRS_project/Cross-disorder/Output
lls
cd igmm/datastore/GenScotDepression/users/amelia/PRS_project/Cross-disorder/Output
ls
put -R significant_snps/

exit
exit
```

#### Submitting jobs on Eddie
What is a job? 
What does this mean?
How do you submit a job?

1. Create a job file and edit it. eg:
```
touch jobfile.sh
vi jobfile.sh # See below for tips on using vi editor.
```
My scripts usually contain the following:
```
#!/bin/sh
#$ -N job_name
#$ -e path/to/where/I/want/error/log/to/be/stored
#$ -o path/to/where/I/want/output/log/to/be/stored
#$ -l h_vmem=8G
#$ -l shared_mem = 4
#$ -m beas
#$ -M emailaddress@ed.ac.uk

# Initialise the environment modules
. /etc/profile.d/modules.sh
# Load R
module load igmm/apps/R/3.6.1
# Run the R script
Rscript --vanilla /exports/igmm/eddie/GenScotDepression/amelia/test.R
```
This will run my R script called `test.R`
`R CMD BATCH` can be used instead of `Rscript`. It's important you load the module for whatever package you want to use, which includes initialising the environment modules.

2. Submit the job to Eddie.
```
qsub jobfile.sh
```
For more info on the qsub command see the man page. The flags denoted in the job file can also be passed in the command itself. But ensure you put them before the name of the job file you want to submit. 
eg:
```
qsub -N job_name -l h_vmem = 8G jobfile.sh
```

3. Check status of your job using:
```
qstat
```
Returns the job id (which you should make a note of in case you need to troubleshoot later) and other useful things, like whether your job is queuing `qw` or running `r`. Jobs with larger memory will queue for longer.

##### Troubleshooting when your job doesn't work.
This is why I like to use the `-m` and `-M` flags for qsub to send me an email when the job starts, finishes or gets an error and aborts. As if it gets an error and doesn't run you can look at this email to find out why. Alternatively you can inspect your log files. Or try `qacct -j <jobid>` where `<jobid` is the number of your job, you'll have to make a note of this when your job is running by using the command `qstat`.

If it says `EXIT STATUS = 137` and ` died through signal KILL (9)` then the job didn't have enough memory to run. There's not really an easy way to tell how much memory your job will need. Couple of things to try... 1) If it's a massive job, eg. GWAS, see if you can submit it in chunks as smaller jobs (see below for an example of when I had to do this), 2) Otherwise I will start with submitting the job with 1G memory and then keep doubling this until the job runs.

Other things to try... usually I would start by writing and running code on an interactive login node to just check everything is running and then submit it as a job. I would also go onto the interactive login node to troubleshoot why a job hasn't run after looking at the log files for some clues.
This logs us into an interactive node on eddie and opens R: **Really important you do this before running R etc., don't run R or do any heavy memory stuff on the login node that you are automatically on when you ssh into eddie.**
```
qlogin -l h_vmem=32G 
. /etc/profile.d/modules.sh
module load igmm/apps/R/3.6.1
R
```
**It's also really important to exit from this interactive login node when you are finished to free up space for other users.** To do this just type `exit` after you are finished. If your computer crashes or you get that annoying "broken pipe" error message because you lost connection through the VPN then this login node will still be active. It's important to log back into eddie and do: `qstat` and then `qdel -j <job_id>` where `<jobid>` is the id of the `QLOGIN` job.



________________

#### Advanced tips:
##### Setting up an SSH key
This is more secure than a password and means you don't have to keep typing your password everytime you log onto Eddie.
On your local computer:
```
# Check you have an .ssh directory, else make one using `mkdir`.
# Generate an SSH key
ssh-keygen -t rsa
# Copy ssh public key and put them onto eddie
ssh-copy-id -i ~/.ssh/id_rsa.pub s1211670@eddie.ecdf.ed.ac.uk
# So you don't have to keep entering your password:
ssh-agent $BASH
ssh-add ~/.ssh/id_rsa
# Edit your ~/.ssh/config file
vi ~/.ssh/config
# Paste the following into it:
HostName s1211670@eddie.ecdf.ed.ac.uk
RemoteForward 52698 localhost:52698
# exit vi editor
# check if it worked (you shouldn't have to enter your password now)
ssh s1211670@eddie.ecdf.ed.ac.uk
```
There's probably better guides to do this if you have a google. Please paste any good resources in the ["Useful links section".](#useful-links)

##### Getting variables from your bash environment into an R script
I found this really useful when I wanted to change my script to be submitted as multiple jobs rather than just one. This is the general concept:

* The jobs I ran were as follows:
    ```
    qsub -N GWAS_1 -v x=1 -v y=300000 job_GWAS.sh
    qsub -N GWAS_2 -v x=300001 -v y=600000 job_GWAS.sh
    qsub -N GWAS_3 -v x=600001 -v y=900000 job_GWAS.sh
    ```
    Where I wanted to run the same job_GWAS.sh script but with different variables submitted to R. ie the x and y variables (which in this case specified different rows of a dataframe). The `-v` flag specifies these variables which will now exist in the bash environment that the job script is run in. You need a new `-v` flag for each new variable.

* My job script looked like this:
    ```
    #!/bin/sh
    #$ -e /exports/igmm/eddie/GenScotDepression/amelia/PRS_project/Cross-disorder/job_logs
    #$ -o /exports/igmm/eddie/GenScotDepression/amelia/PRS_project/Cross-disorder/job_logs
    #$ -l h_vmem=4G
    #$ -m beas
    #$ -M edmondson.staita@gmail.com

    # Initialise the environment modules
    . /etc/profile.d/modules.sh
    module load igmm/apps/R/3.6.1

    Rscript --vanilla /exports/igmm/eddie/GenScotDepression/amelia/PRS_project/Cross-disorder/Scripts/common_factorGWAS.R ${x} ${y}
    ```
    The important bits here are `${x}` and `${y}`, and the fact these are listed after I specified the R script to run. The order you list them here are the order they are read in R (which is important).

* My R script contained the code:
    ```
    ##First read in the arguments listed at the command line in bash
    args <- commandArgs(trailingOnly=TRUE) # This is the amazing command that reads variables stored in your bash environment into R.
    # Separate them out using their index (this is a bit error prone as if you mess up the order or there are more than 2 bash variables you'll get the wrong input, there's probably more elegant ways to do this).
    x <- as.numeric(args[[1]])
    y <- as.numeric(args[[2]])

    # Run the common factor GWAS on a subset of SNPs (from row x to row y of my p_sumstas dataframe).
    # x and y are specifed by the bash job script.
    pfactor <- commonfactorGWAS(covstruc = LDSCoutput, SNPs = p_sumstats[x:y,], estimation = "DWLS", cores = NULL, toler = FALSE, SNPSE = FALSE, parallel = FALSE, Output = NULL,GC="standard",MPI=FALSE)
    ```

##### Using Sublime text editor from Eddie by creating an ssh tunnel (yes, it's as exciting as it sounds!)
Fed up of using that annoying vi editor? Yes, I was too. So I found out how to open my text to edit in Sublime instead. I'm sure there's a way to do this with other text editors, please share details below if you know how.

```
# On your local machine find your .ssh folder and edit the config file. eg.:
vi ~/.ssh/config
# Insert the following: (press "i" to edit in vi)
HostName s1211670@eddie.ecdf.ed.ac.uk # Change to your UUN
RemoteForward 52698 127.0.0.1:52698
# exit vi (esc and then :wq)
```
Install Sublime on your local machine. Then install rsub package on your local Sublime (using Sublime's package manager).
<br/>
Then go back to your bash terminal on your local computer and log into eddie using:
```
# Install rmate on Eddie (this talks to rsub through the SHH tunnel)
ssh -R 52698:localhost:52698 s1211670@eddie.ecdf.ed.ac.uk

wget -O /exports/igmm/eddie/GenScotDepression/amelia/packages/rsub \https://raw.github.com/aurora/rmate/master/rmate
chmod a+x /exports/igmm/eddie/GenScotDepression/amelia/packages/rsub #Changes permissions of folder.

# Edit your bash profile
vi ~/.bash_profile
# Add path to rsub command by copying and pasting this:
export PATH=$PATH:/exports/igmm/eddie/GenScotDepression/amelia/packages/
# exit vi 

# Opens text file in sublime.
cd /exports/igmm/eddie/GenScotDepression/amelia/PRS_project/Cross-disorder/Scripts/
rsub common_factor.R
```

##### Editing your bash profile, (and then accidentally breaking it so eddie wouldn't open any more, oops)
I once edited my bash profile with something that didn't work and then couldn't log into eddie! Luckily I could do
```
ssh -t s1211670@eddie.ecdf.ed.ac.uk /bin/sh
vi ~/.bash_profile
```
and remove the broken code.


___________________
#### Some general useful Linux commands

* `cd path/to/directory/you/want/to/change/to` Change directory.
* `cd ../` Change directory by moving back one folder.
* `pwd` Print your working directory.
* `vi filename` Open a file in vi text editor (useful to have a quick look and edit a small file).
    - Press "i" on your keyboard to insert text.
    - Press "esc" when you have finished, followed by ":wq", the "w" means "write" and the "q" means "quit".
* `less filename` View contents of file. Press "q" when finished to return to command line.
* `cat` Prints contents of file in terminal. Not so good for really large files!
* Saving bash variables and printing them in the terminal, eg.:
    ```
    $ x=2
    $ echo ${x}
    2
    ```


_____________
#### Useful links
* 






